src/mpid/ch4/shm/ipc/xpmem/xpmem_control.c

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpidimpl.h"
+#include "xpmem_pre.h"
+#include "xpmem_recv.h"
+#include "xpmem_control.h"
+
+int MPIDI_IPC_xpmem_send_lmt_coop_recv_fin_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
```

Comment:
Why does xpmem need to do all this?  It should only be worried about opening and mapping memory buffers, isn't it?

Reply:
We think the coop protocol is XPMEM-custom protocol for now. It is unclear whether it benefits GPU. When it is ready to be promoted to a IPC generic protocol, we can move them into IPC layer. It also simplifies this PR.

Reply:
I think @minsii idea here was to have specialization of functionality in the netmod. Since cooperative copy is only needed for XPMEM and not, e.g., for GPU. I guess if more cases of coop copy become available we can perhaps move this to the IPC layer? @minsii what do you think?

Reply:
@giordano We responded at the same time :-) Please see my comment above.

Reply:
There's nothing XPMEM-specific about the coop protocol.  It's a mistake to put it in there.  FYI, GPUs can use it too.  It'll use two DMA engines instead of one.